File: libUtil/src/main/java/org/waynezhou/libUtil/event/EventGroup.java
package org.waynezhou.libUtil.event;


public abstract class EventGroup <TEventGroup extends EventGroup<TEventGroup>>  {

    protected EventGroup() {

    }

    public <TEventArgs> TEventGroup on(EventGroupHolderSelector<TEventGroup, TEventArgs> selector, EventListener<TEventArgs> listener){
        selector.select((TEventGroup)this).addOnListener(listener);
        return (TEventGroup)this;
    }

    public <TEventArgs> TEventGroup once(EventGroupHolderSelector<TEventGroup, TEventArgs> selector, EventListener<TEventArgs> listener){
        selector.select((TEventGroup)this).addOnceListener(listener);
        return (TEventGroup)this;
    }

    public <TEventArgs> EventHolder<TEventArgs>.ListenerToken on_(EventGroupHolderSelector<TEventGroup, TEventArgs> selector, EventListener<TEventArgs> listener){
        return selector.select((TEventGroup)this).addOnListener(listener);
    }

    public <TEventArgs> EventHolder<TEventArgs>.ListenerToken once_(EventGroupHolderSelector<TEventGroup, TEventArgs> selector, EventListener<TEventArgs> listener){
        return selector.select((TEventGroup)this).addOnceListener(listener);
    }

    private EventGroup<TEventGroup>.Invoker invoker;
    protected EventGroup<TEventGroup>.Invoker getInvoker(){
        if(invoker==null) invoker = new Invoker((TEventGroup)this);
        return invoker;
    }

    public <TEventArgs> TEventGroup remove(EventGroupHolderSelector<TEventGroup, TEventArgs> selector, EventHolder<TEventArgs>.ListenerToken token){
        selector.select((TEventGroup)this).removeListener(token);
        return (TEventGroup)this;
    }
    public <TEventArgs> TEventGroup remove(EventGroupHolderSelector<TEventGroup, TEventArgs> selector, EventListener<TEventArgs> listener){
        selector.select((TEventGroup)this).removeListener(listener);
        return (TEventGroup)this;
    }

    public TEventGroup apply(EventGroupApplier<TEventGroup> applier){
        TEventGroup group = (TEventGroup)this;
        applier.apply(group);
        return group;
    }


    public class Invoker{
        private final TEventGroup  group;
        private Invoker(TEventGroup group){
            this.group = group;
        }
        public <TEventArgs> void invoke(EventGroupHolderSelector<TEventGroup, TEventArgs> selector, TEventArgs eventArgs){
            selector.select(group).invoke(eventArgs);
        }
    }

    @FunctionalInterface
    public interface EventGroupApplier<TEventGroup extends EventGroup<TEventGroup>> {
        void apply(TEventGroup pack);
    }

    @FunctionalInterface
    public interface EventGroupHolderSelector<TEventGroup extends EventGroup<TEventGroup>, TEventArgs>{
        EventHolder<TEventArgs> select(TEventGroup group);
    }
}

File: libUtilKt/src/main/java/org/waynezhou/libutilkt/event/EventGroup.kt
package org.waynezhou.libutilkt.event

abstract class EventGroup<TEventGroup : EventGroup<TEventGroup>> protected constructor() {

    @Suppress("UNCHECKED_CAST")
    fun <TEventArgs> on(
        selector: (TEventGroup)->EventHolder<TEventArgs>,
        listener: EventListener<TEventArgs>
    ): TEventGroup {
        selector( this as TEventGroup).addOnListener(listener)
        return this
    }

    @Suppress("UNCHECKED_CAST")
    fun <TEventArgs> once(
        selector: (TEventGroup)->EventHolder<TEventArgs>,
        listener: EventListener<TEventArgs>
    ): TEventGroup {
        selector(this as TEventGroup).addOnceListener(listener)
        return this
    }


    @Suppress("UNCHECKED_CAST")
    fun <TEventArgs> on_(
        selector: (TEventGroup)->EventHolder<TEventArgs>,
        listener: EventListener<TEventArgs>
    ): EventHolder<TEventArgs>.ListenerToken {
        return selector(this as TEventGroup).addOnListener(listener)
    }

    @Suppress("UNCHECKED_CAST")
    fun <TEventArgs> once_(
        selector: (TEventGroup)->EventHolder<TEventArgs>,
        listener: EventListener<TEventArgs>
    ): EventHolder<TEventArgs>.ListenerToken {
        return selector(this as TEventGroup).addOnceListener(listener)
    }

    inner class Invoker internal constructor(
        private val group: TEventGroup)
    {
        operator fun <TEventArgs> invoke(
            selector: (TEventGroup)->EventHolder<TEventArgs>,
            eventArgs: TEventArgs
        ) {
            selector(group).invoke(eventArgs)
        }
    }
    private var invoker: EventGroup<TEventGroup>.Invoker? = null
    @Suppress("UNCHECKED_CAST")
    protected fun getInvoker(): EventGroup<TEventGroup>.Invoker {
        if (invoker == null) invoker = Invoker(this as TEventGroup)
        return invoker!!
    }

}

class EventHolder<TEventArgs> {
    private inner class Listener(
        val isOnce: Boolean,
        val listener: EventListener<TEventArgs>
    )
    private val listeners: MutableList<Listener> = mutableListOf()
    internal fun addOnceListener(listener: EventListener<TEventArgs>): ListenerToken {
        val l = Listener(true, listener)
        val token = ListenerToken(l.hashCode())
        listeners.add(l)
        return token
    }

    internal fun addOnListener(listener: EventListener<TEventArgs>): ListenerToken {
        val l = Listener(false, listener)
        val token = ListenerToken(l.hashCode())
        listeners.add(l)
        return token
    }

    internal fun removeListener(listener: EventListener<TEventArgs>) {
        listeners.removeIf{ x: Listener -> x.listener === listener }
    }

    internal fun removeListener(token: ListenerToken) {
        listeners.removeIf { x: Listener -> x.hashCode() == token.hashcode }
    }

    internal fun invoke(e: TEventArgs) {
        val removedIdx: MutableList<Int> = mutableListOf()
        for ((idx, listener) in listeners.withIndex()) {
            listener.listener.invoke(e)
            if (listener.isOnce) {
                removedIdx.add(idx)
            }
        }
        removedIdx.forEach{listeners.removeAt(it)}
    }

    inner class ListenerToken(val hashcode: Int)
}

fun interface EventListener<TEventArgs> {
    operator fun invoke(e: TEventArgs)
}